CXINT-1295 - CDC B2B- MANAGE USERS

anjana-bl opened the pull request:
Under 'my company'->'users', when CDC & B2B is true,
1. Add button to be replaced with Manage Users button. Clicking on this button should open a Gigya portal in a new tab.
2. In user details screen, Edit, Disable, Change Password and hyperlink under Units should be hidden.

This PR is a follow-up of https://github.com/SAP/spartacus/pull/16748  

The following is a checklist for new epics or features acceptance for Spartacus, based on [our Definition of Done document](https://sap.github.io/spartacus-docs/definition-of-done/)

## General

- [x] Epic has been tested on a production-like server using Spartacus installation script
- [x] Epic has been tested in all supported browsers and devices (Check list on DoD page)
- [x] Epic has sufficient end to end (cypress) test coverage
- [x] Schematics code deprecations have been provided and tested
- [x] Epic complies with [Security best practices](https://sap.github.io/spartacus-docs/security-best-practices/)
- [x] Epic complies with [Accessibility best practices](https://sap.github.io/spartacus-docs/a11y-best-practices/)
- [x] Epic CSS supports directionality (if required)
- [ ] Epic code scans have been conducted [Compliance scans for branches](https://wiki.one.int.sap/wiki/display/spar/Compliance+scans+for+branches)
- [x] Epic renders successfully in SSR

## Audits/reviews

- [x] Architecture review of the epic
- [x] Security and threat modeling audit/review from the Security team
- [x] Accessibility audit/review from the UX team

## New Library

If your epic will be introduced or introduces a new library:

- [x] Library provides schematics to install the library
- [x] Library modules can be lazy loaded appropriately
- [x] Library has been tested for a customer using Spartacus installation script

## Sample data

If the new feature requires new or updated sample data for a specific Commerce backend version:

- [x] Sample data has been added to the sampledata repo and merged to the next version to be released
- [x] New sample data has been loaded and tested on a dev server
- [x] New sample data has been loaded and tested in our CI server

## Documentation

- [x] Documentation input is provided for it to be published on the doc site.

anjana-bl committed:
new feature code + unit tests + snapshot test

anjana-bl committed:
Merge branch 'develop' into cdc-manage-users-2

anjana-bl committed:
adding ng-package.json + fixing sonar errors

anjana-bl committed:
updating unit test

anjana-bl committed:
reverting changes done in core files

anjana-bl committed:
fixing projects/schematics/ unit test

fixing errors when executing "Running unit tests and code coverage for schematics library"

anjana-bl committed:
Merge branch 'develop' into cdc-manage-users-2

SherwinVarghese commented on integration-libs/cdc/manage-users/cdc-user-list.service.ts:
```diff
@@ -0,0 +1,70 @@
+  protected subscription: Subscription = new Subscription();
+
+  onCreateButtonClick(): void {
+    this.cdcJsService.didLoad().subscribe((cdcLoaded) => {
```
`this.cdcJsService.didLoad()` check can be skipped, since the user would have logged in already using CDC SDK.

anjana-bl replied:
done

anjana-bl committed:
adding e2e test case

anjana-bl committed:
Formating

anjana-bl committed:
Merge branch 'develop' into cdc-manage-users-2

Platonn commented on feature-libs/organization/administration/components/shared/list/list.service.ts:
```diff
@@ -158,4 +158,24 @@ export abstract class ListService<T, P = PaginationModel> {
+  /**
+   * This method will return the label for create button
+   */
+  getCreateButtonLabel(): string {
+    return 'organization.add';
```
Let's change the return type to `Translatable`, so to be obvious that this is a translation label.

```ts
getCreateButtonLabel(): Translatable {
   return { key: 'organization.add' }
}

anjana-bl replied:
done

Platonn commented on feature-libs/organization/administration/components/shared/list/list.service.ts:
```diff
@@ -158,4 +158,24 @@ export abstract class ListService<T, P = PaginationModel> {
     ...args: any
   ): Observable<EntitiesModel<T> | undefined>;
+
+  /**
+   * If this method returns true, hyperlink will be shown to add new user.
+   * If this method returns false, button will be displayed.
+   */
+  showLink(): boolean {
+    return true;
+  }
```
As @armanozak mentioned, let's not return a boolean (and let's change the method name), but something more generic like `enum` value (so to be prepared for possible new extensions in the future, if only needed).

e.g.
```ts
export enum CreateButtonType {
  LINK = 'LINK';
  BUTTON = 'BUTTON';
}
```

and
```ts
getCreateButtonType(): CreateButtonType {
  return CreateButtonType.LINK;
} 
```

Please also change accordingly the logic/template of the generic `ListComponent` and your `CdcListService`

PS. names of enums and methods are open for discussion

anjana-bl replied:
done

Platonn commented on feature-libs/organization/administration/components/shared/list/list.service.ts:
```diff
@@ -158,4 +158,24 @@ export abstract class ListService<T, P = PaginationModel> {
+  }
+
+  /**
+   * This method will be called when the button to create new user is clicked.
```
Let's replace the word `user` with `item` in this code comment. This service is generic - used not only for Users, but also for Budgets, CostCenters, etc.

anjana-bl replied:
done

Platonn commented on feature-libs/organization/administration/core/guards/admin.guard.ts:
```diff
@@ -15,17 +15,33 @@ import {
   ) {}
 
+  canMatch(): boolean {
+    const promise = this.b2bUserService.isUpdatingUserAllowed();
+    if (!promise) {
```
Nitpick: Calling a variable `promise` is misleading. It's not a `Promise`. Would you mind changing it to `const isUpdatingUserAllowed`?

anjana-bl replied:
done

Platonn commented on feature-libs/organization/administration/core/guards/admin.guard.ts:
```diff
@@ -15,17 +15,33 @@ import {
+  canMatch(): boolean {
+    const promise = this.b2bUserService.isUpdatingUserAllowed();
+    if (!promise) {
+      this.routingService.go({ cxRoute: 'organization' });
+      this.globalMessageService.add(
+        { key: 'organization.notification.notExist' },
+        GlobalMessageType.MSG_TYPE_WARNING
+      );
+    }
+    return promise;
```
When redirecting to another URL inside a guard, it's recommended to return the `UrlTree` of the redirected URL, instead of returning `false` and triggering a new navigation via `.go()`

```ts
    const isUpdatingUserAllowed = this.b2bUserService.isUpdatingUserAllowed();
    if (isUpdatingUserAllowed) {
      this.globalMessageService.add(
        { key: 'organization.notification.notExist' },
        GlobalMessageType.MSG_TYPE_WARNING
      );
      return this.router.parseUrl(
        this.semanticPathService.get('organization') ?? ''
      );
    }
    return true;
```

PS 1. for more on returing `UrlTree` from a guard, see this article https://juristr.com/blog/2018/11/better-route-guard-redirects/

PS 2. I can see that in a few guards in the `organization` library we use `.go()` in redirects, but it's an antipattern and should not be copied. In other guards in Spartacus we return from guards a proper `UrlTree` in case of redirect.

anjana-bl replied:
done

Platonn commented on feature-libs/organization/administration/core/guards/admin.guard.ts:
```diff
@@ -15,17 +15,33 @@ import {
+    protected location: Location
   ) {}
 
+  canMatch(): boolean {
```
I suggest not to enlarge the scope of the generic `AdminGuard`, which used to be responsible only for protecting pages accessible only for the Admin's role. This guard can be introduced on any page, not only User management pages (but also budgets, cost centers, etc) - so let's not introduce here a hard dependency on `b2bUserService`.

 Instead I suggest to create a brand new guard for your custom use case (which should be scoped only to User btw.)

Platonn commented on integration-libs/cdc/manage-users/manage-users.module.ts:
```diff
@@ -0,0 +1,21 @@
+    { provide: UserListService, useClass: CdcUserListService },
+  ],
+})
+export class ManageUsersModule {}
```
What do you think about renaming the phrase "ManageUsers" to "CdcOrganizationAdministration"? (to follow similar naming convetions as we took when extending sub-libraries UserProfile and UserAccount - to derive the Cdc module from the original module that we extend)

Since we're extending the original entry point `@spartacus/organization/administration`, I suggest rename this module rather to `CdcOrganizationAdministration` (and accordingly: filename `cdc-organization-administration.module` and constants in schematics `CDC_ORGANIZATION_ADMINISTRATION_MODULE`). Also let's rename the CDC library's sub-entry-point from `@spartacus/cdc/manage-users` to `@spartacus/cdc/organization-administration`.

PS. Another benefit of this renaming is that we can put here possibly more extensions of the original OrganizationAdministration module - e.g. disabling the default UI parts in routes of UserGroups and Units - not to allow for modifying an user from those views - as we've recently discussed offline.

anjana-bl replied:
done

Platonn commented:
We've discussed if offline, but let me drop here a memo:

I've checked in the default B2B Spartacus, that the organization’s admin can change user’s roles and rights in the Units view. See URL http://localhost:4200/powertools-spa/en/USD/organization/units/Rustic%20Retail/users/9a6de44d-23da-4a56-9c81-f1a103a5388f/roles
and also
I’ve also checked that an User can be unassigned from an UserGroup (in UserGroups view). See URL:
http://localhost:4209/powertools-spa/en/USD/organization/user-groups/limitedPermissions/users

(to run default B2B MyCompany feature in Spartacus, please open develop branch, run `SPA_ENV=b2b yarn start` and then login e.g. as `linda.wolf@rustic-how.com` inside the opened B2B Storefront)

---

Screenshot of: http://localhost:4200/powertools-spa/en/USD/organization/units/Rustic%20Retail/users/9a6de44d-23da-4a56-9c81-f1a103a5388f/roles
![image](https://user-images.githubusercontent.com/4001059/218599194-2711a468-134b-4ce3-b0f7-667b648709b3.png)

----

Screenshot of: http://localhost:4209/powertools-spa/en/USD/organization/user-groups/limitedPermissions/users
![image](https://user-images.githubusercontent.com/4001059/218599209-6f793b65-3e2c-4d4f-96dd-46df22ef2258.png)

Platonn commented on feature-libs/organization/administration/core/guards/admin.guard.ts:
```diff
@@ -15,17 +15,33 @@ import {
-    protected globalMessageService: GlobalMessageService
+    protected globalMessageService: GlobalMessageService,
+    protected b2bUserService: B2BUserService,
+    protected location: Location
```
Nitpick: this service is not used

Platonn commented on integration-libs/cdc/manage-users/cdc-user-list.service.ts:
```diff
@@ -0,0 +1,70 @@
+  ) {
+    super(tableService, userService);
+  }
+  protected subscription: Subscription = new Subscription();
```
Nitpick: let's remove an unused property

anjana-bl replied:
this is being used now (based on the below comment)

Platonn commented on integration-libs/cdc/manage-users/cdc-user-list.service.ts:
```diff
@@ -0,0 +1,70 @@
+  protected subscription: Subscription = new Subscription();
+
+  onCreateButtonClick(): void {
+    this.cdcJsService.didLoad().subscribe((cdcLoaded) => {
+      if (cdcLoaded) {
+        // get current organization ID using CDC Gigya SDK
+        this.cdcJsService.getOrganizationContext().subscribe({
+          next: (response) => {
```
Instaed of doing inner subscriptions, which are harder to read, lets simplify this code, by using: `pipe` and `switchMap`
```ts
  const sub = this.cdcJsService.didLoad().pipe(
    switchMap(cdcLoaded => {
       /* ... */
       return this.cdcJsService.getOrganizationContext()
    }),
    tap(response => {
      /* ... */
        this.cdcJsService.openDelegatedAdminLogin(response.orgId);
      /* ... */ 
    })
  ).subscribe(); // we subscribe only once here

  // moreover, let's make sure to cleanup subscriptions in case when an user leaves (destroys) this Component but the observable logic would not ever complete:
  this.subscription.add(sub);


/* ... */
// and eventually on component's destroy let's make sure to dispose any possibly pending subscriptions
ngOnDestroy(){
  this.subscription?.unsubscribe();
}  
```

anjana-bl replied:
done

anjana-bl committed:
changes for review comments

anjana-bl committed:
Merge branch 'develop' into cdc-manage-users-2

anjana-bl committed:
Update cdc-administration.module.ts

anjana-bl committed:
Merge branch 'cdc-manage-users-2' of https://github.com/SAP/spartacus into cdc-manage-users-2

anjana-bl commented:
Hi Kris, I had discussion with @SherwinVarghese .. so below are the points which we finalised.
> 'Create' and 'Roles and Rights' links under 'All Units->Unit Details->Users' will be hidden. 
> No changes are done from 'User Groups' perspective, as this is independent of CDC.

Platonn commented on feature-libs/organization/administration/components/shared/list/list.component.html:
```diff
@@ -51,14 +51,16 @@ <h3>
-          >
-            {{ 'organization.add' | cxTranslate }}
-          </a>
+          <ng-container *ngIf="createButtonType === 'LINK'; else showButton">
```
For type safety, can we use here `CreateButtonType.LINK` instead of hardcoded string `'LINK'`, please?

you might need to assign the enum as a public field in the TS class (so it can be "seen" in the HTML):
```ts
import {CreateButtonType} from '......';

export class ListComponent {
   CreateButtonType = CreateButtonType; // <-----
   // ...
}
```

anjana-bl replied:
done

Platonn commented on integration-libs/cdc/root/service/cdc-js.service.ts:
```diff
@@ -265,6 +266,39 @@ export class CdcJsService implements OnDestroy {
+   *
+   */
+  getOrganizationContext(): Observable<{ orgId: string }> {
+    return new Observable<{ orgId: string }>((gotContext) => {
```
Nitpick: For having the correct semantics of the `new Observable`'s function argument, can we rename the `gotContext` to `subscriber`, please?

The argument is a reference to the subscriber observable, not an information whether we got context or not. For more see https://rxjs.dev/guide/observable

PS. I can see in other places we missed that in the past, not calling the argument properly `subscriber`

anjana-bl replied:
done

Platonn commented on projects/storefrontapp-e2e-cypress/cypress/helpers/vendor/cdc/cdc.ts:
```diff
@@ -16,6 +16,13 @@ export function registerUser() {
 const nativeUser = getSampleUser();
 
+export const cdcB2BDelegateAdminUser = {
+  userId: '1002',
+  fullName: 'Anjana BL',
+  email: 'anjana.b.l+1@sap.com',
+  password: 'Anjana1201!',
+};
```
Question: is this a special user that you've created in the sample data via Impex? I'm asking to make sure this user will be preserved in the sample data, even if we'll reinitialize in the future the OCC server (with CDC-flavoured sample data).

anjana-bl replied:
this user is persisted in CDC always, so there is no need to create this user in each commerce server. even if the commerce server is reinitialised, this user will get automatically created. There is no need for separate sample data to be imported everytime.

Platonn commented on projects/storefrontapp-e2e-cypress/cypress/integration/vendor/cdc/b2b/b2b-manage-users-e2e-spec.ts:
```diff
@@ -0,0 +1,59 @@
+  beforeEach(() => {
+    cy.visit('/powertools-spa/en/USD/login');
+    loginUser(cdcB2BDelegateAdminUser);
+    cy.wait(3000);
```
Waiting a hardcoded amount of time (e.g. 3000ms) is not recommended. Instead we should await e.g. specific the background http request.

Here's an example in Spartacus codebase: [usage](https://github.com/SAP/spartacus/blob/3fe8b68c0899d822a983879abc95fa5ee28d351c/projects/storefrontapp-e2e-cypress/cypress/integration/vendor/epd-visualization/visualization-lookup/visualization-lookup.e2e-spec.ts#L26) and [definition](https://github.com/SAP/spartacus/blob/3fe8b68c0899d822a983879abc95fa5ee28d351c/projects/storefrontapp-e2e-cypress/cypress/integration/vendor/epd-visualization/visualization-lookup/visualization-lookup.e2e-spec.ts#L8-L13) of " Cypress route alias".
  
For more, see:
- Cypress best practices for 'wait': https://docs.cypress.io/guides/references/best-practices#Unnecessary-Waiting
- Cypress docs for 'wait: https://docs.cypress.io/api/commands/wait

--
This remark applies also to other places in this file

anjana-bl replied:
done

anjana-bl committed:
code review changes

anjana-bl committed:
code review changes

anjana-bl committed:
changes to e2e testcase

SherwinVarghese commented on projects/storefrontapp-e2e-cypress/cypress/integration/vendor/cdc/b2b/b2b-manage-users-e2e-spec.ts:
```diff
@@ -0,0 +1,66 @@
+import { loginUser } from '../../../../helpers/checkout-flow';
```
@anjana-bl as discussed, please rename this file as `projects/storefrontapp-e2e-cypress/cypress/integration/vendor/cdc/b2b/b2b-scenarios.core-e2e-spec.ts`

anjana-bl replied:
done

anjana-bl committed:
changing e2e file name

anjana-bl committed:
Merge branch 'develop' into cdc-manage-users-2

anjana-bl committed:
Update cdc-administration.module.ts

anjana-bl committed:
Update ng-package.json

armanozak committed:
chore: change the order in build:libs script (#16910)

armanozak committed:
AdministrationWrapperModule test & sample (#16911)

* test: add test for administration wrapper module

* chore: add CdcAdministrationModule to sample app

anjana-bl committed:
moving extension loading into feature module from wrapper module

armanozak committed:
Override ManageUsersListComponent config (#16921)

anjana-bl committed:
changes to load cdc module in organization wrapper module

anjana-bl committed:
Merge branch 'develop' into cdc-manage-users-2

anjana-bl committed:
Update package.json

anjana-bl committed:
Update cdc-administration.module.ts

anjana-bl committed:
Merge branch 'develop' into cdc-manage-users-2

anjana-bl committed:
Merge branch 'develop' into cdc-manage-users-2

anjana-bl committed:
Update package.json

triggering a push to update https://github.tools.sap/cx-commerce-storefronts/composable-storefront/tree/cdc-manage-users-2

Platonn commented on integration-libs/cdc/organization/administration/cdc-administration.module.ts:
```diff
@@ -0,0 +1,55 @@
+
+@NgModule({
+  providers: [
+    //to override UserListService in ListComponent
+    provideDefaultConfig(<CmsConfig>{
+      cmsComponents: {
+        ManageUsersListComponent: {
+          providers: [
+            {
+              provide: UserListService,
+              useExisting: CdcUserListService,
+            },
+            userCmsConfig.cmsComponents?.ManageUsersListComponent?.providers ||
+              [],
+          ],
+        },
+      },
+    }),
+    //to override B2BUserService in UserDetailsComponent, UnitUserListComponent
+    { provide: B2BUserService, useClass: CdcB2BUserService },
+    //to override B2BUserService in UnitUserRolesCellComponent
+    provideDefaultConfig(<CmsConfig>{
+      cmsComponents: {
+        ManageUnitsListComponent: {
+          providers: [
+            {
+              provide: B2BUserService,
+              useExisting: CdcB2BUserService,
+            },
+            unitsCmsConfig.cmsComponents?.ManageUnitsListComponent?.providers ||
+              [],
+          ],
+        },
+      },
+    }),
```
Is this all _really_ neeed? Can't we replace it with just 2 lines providers in `@NgModule.providers`:
```ts
{ provide: UserListService, useExisting: CdcUserListService },
{ provide: B2BUserService, useExisting: CdcB2BUserService },
```
?

---
Reason:
I can see `UserListService` is already `providedIn: root`: https://github.com/SAP/spartacus/blob/bd5c7c93a2aa46f30600ee3d3fd09835beebe6bf/feature-libs/organization/administration/components/user/services/user-list.service.ts#L37-L40

So I guess we don't need to over-provide it on the cms-mapping level, but it should suffice to over-provide it just on the level of `@NgModule.

Thanks to that, we could avoid potentially unnecessary referencing `userCmsConfig.cmsComponents?.ManageUsersListComponent?.providers` and merging it with our custom over-provider on the cms-mapping level.

and Analogically for `B2BUserService`.

anjana-bl replied:
Hi Kris, making the suggested change is not overriding B2BUserService with CdcB2BUserService in UnitUserRolesCellComponent and UserListService with CdcUserListService in ListComponent.

anjana-bl committed:
Merge branch 'develop' into cdc-manage-users-2
